### PR TITLE
feat(gatsby): Gatsby Head API

### PR DESCRIPTION
**Pull request by marvinjude**
## Description

[Link to RFC](https://github.com/gatsbyjs/gatsby/discussions/35841)

### Todo

- [x] Warnings and descriptive error messages
- [x] End-to-end tests - https://github.com/gatsbyjs/gatsby/pull/35979
- [x] HMR for Head Export -  https://github.com/gatsbyjs/gatsby/pull/36060
- [x]  Make Head export work in `tsx` pages - https://github.com/gatsbyjs/gatsby/pull/36114
- [x]  More e2e tests (tags cleanup, `.tsx` pages)

### Documentation

https://github.com/gatsbyjs/gatsby/pull/36125

### CODE REVIEW

File: packages/gatsby/cache-dir/dev-loader.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ class DevLoader extends BaseLoader {
 
       return (
         this.asyncRequires.components[chunkName]()
-          .then(preferDefault)
+          .then(component => component)
```
Comment (pieh) [inline packages/gatsby/cache-dir/dev-loader.js:43]:
```suggestion
```

**Comment by marvinjude** [comment]
Notes 
- Add tests for Fs Route
-  Pass only needed props (data, pageContext, params), Other props like `pageResources` may be removed in v5

File: packages/gatsby/cache-dir/loader.js [hunk 2]
```diff
@@ -577,7 +577,7 @@ export class ProdLoader extends BaseLoader {
 
       return (
         asyncRequires.components[chunkName]()
-          .then(preferDefault)
+          .then(component => component)
```
Comment (pieh) [inline packages/gatsby/cache-dir/loader.js:580]:
```suggestion
```

File: packages/gatsby/cache-dir/page-renderer.js [hunk 3]
```diff
@@ -1,35 +1,115 @@
...
-      ({ result }) => {
-        return { element: result, props }
+      key: props.path || props.pageResources.page.path,
+    }
+  )
+
+  const pageComponent = props.pageResources.component
+
+  if (pageComponent.head) {
+    if (typeof pageComponent.head !== `function`)
```
Comment (marvinjude) [inline packages/gatsby/cache-dir/page-renderer.js:52]: Allow check to support class components

File: packages/gatsby/cache-dir/page-renderer.js [hunk 4]
```diff
@@ -1,35 +1,115 @@
...
+function Caller({ children, callback }) {
+  useEffect(() => {
+    callback()
+  }, [callback])
+
+  return children
+}
+
+// TODO: We do something similar in production-app, we  should use a take this function, put it in some utils file and share
+function render(el, Component) {
```
Comment (marvinjude) [inline packages/gatsby/cache-dir/page-renderer.js:19]: TODO: We do something similar in production-app, we should use a take this function, put it in some utils file and share

File: packages/gatsby/cache-dir/page-renderer.js [hunk 5]
```diff
@@ -1,35 +1,115 @@
...
+
+    useEffect(() => {
+      const hiddenRoot = document.createElement(`div`)
+
+      const callback = () => {
+        // Remove previous head nodes
+        const prevHeadNodes = [
+          ...document.querySelectorAll(`[data-gatsby-head]`),
+        ]
+        prevHeadNodes.forEach(e => e.remove())
```
Comment (marvinjude) [inline packages/gatsby/cache-dir/page-renderer.js:71]: Optionally call `.remove`

File: packages/gatsby/cache-dir/static-entry.js [hunk 6]
```diff
@@ -208,12 +210,76 @@ export default async function staticPage({
...
+            <Router baseuri={__BASE_PATH__} component={React.Fragment}>
+              <HeadRouteHandler path="/*" />
+            </Router>
+          </ServerLocation>
+        </StaticQueryContext.Provider>
+      )
+
+      // extract head nodes from string
+      const rawString = renderToString(routerElement)
+      const headNodes = [...parse(rawString).childNodes]
```
Comment (marvinjude) [inline packages/gatsby/cache-dir/static-entry.js:247]: verify if .childNodes is already an array

File: packages/gatsby/cache-dir/static-entry.js [hunk 7]
```diff
@@ -208,12 +210,76 @@ export default async function staticPage({
...
+              rawTagName !== `script`
+                ? `<${rawTagName}> is not a valid head element. Please use one of the following: ${VALID_NODE_NAMES.join(
+                    `, `
+                  )}`
+                : `Do not add scripts here. Please use the <Script> component in your page template instead. For more info see: https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-script/`
+
+            console.warn(warning)
+          }
+        } else {
+          node.setAttribute(`data-gatsby-head`, true)
```
Comment (marvinjude) [inline packages/gatsby/cache-dir/static-entry.js:266]: Remove this

File: packages/gatsby/cache-dir/static-entry.js [hunk 8]
```diff
@@ -208,12 +210,76 @@ export default async function staticPage({
...
+          const element = createElement(
+            rawTagName,
+            {
+              ...attributes,
+              "data-gatsby-head": true,
+            },
+            node.childNodes[0]?.textContent
+          )
+
+          validHeadNodes = [...validHeadNodes, element]
```
Comment (pieh) [inline packages/gatsby/cache-dir/static-entry.js:277]:
```suggestion
          validHeadNodes.push(element)
```

File: packages/gatsby/src/utils/eslint-rules/limited-exports-page-templates.ts [hunk 9]
```diff
@@ -233,6 +233,10 @@ const limitedExports: Rule.RuleModule = {
           return undefined
         }
 
+        if (isApiExport(node, `head`)) {
```
Comment (marvinjude) [inline packages/gatsby/src/utils/eslint-rules/limited-exports-page-templates.ts:236]: Update `isApiExport` not show warming for class head export

File: packages/gatsby/cache-dir/page-renderer.js (hunk 3 above)
Comment (marvinjude) [reply packages/gatsby/cache-dir/page-renderer.js:52]: Check is all fine as class components work right now. We only need to update code so we don't show a warning for class head export

File: packages/gatsby/cache-dir/page-renderer.js (hunk 5 above)
Comment (marvinjude) [reply packages/gatsby/cache-dir/page-renderer.js:62]: We may have fixed the underlying issue that may warrant `e`  not be an element so we may not need to do this until we need to

**Comment by LekoArts** [comment]
Can you revert the image changes for contentful, e.g. `packages/gatsby-source-contentful/src/__tests__/23df7612602d69edfc3eaf89c8c02248/347966-contentful-logo-wordmark-dark (1)-4cd185-original-1582664935 (1).png`

File: packages/gatsby/cache-dir/head/head-export-handler-for-ssr.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      }
+    }
+
+    setHeadComponents(validHeadNodes)
+  }
+}
+
+module.exports = {
+  headHandlerForSSR,
+}
```
Comment (LekoArts) [inline packages/gatsby/cache-dir/head/head-export-handler-for-ssr.js:79]: Why CommonJS export when an `export` in front of `function headHandlerForSSR` would also work?

File: packages/gatsby/cache-dir/fast-refresh-overlay/components/runtime-errors.js [hunk 11]
```diff
@@ -9,12 +9,14 @@ import { Accordion, AccordionItem } from "./accordion"
 function WrappedAccordionItem({ error, open }) {
   const stacktrace = StackTrace.parse(error)
   const codeFrameInformation = getCodeFrameInformation(stacktrace)
-  const filePath = codeFrameInformation?.moduleId
+
+  const modulePath = codeFrameInformation?.moduleId
   const lineNumber = codeFrameInformation?.lineNumber
   const columnNumber = codeFrameInformation?.columnNumber
   const name = codeFrameInformation?.functionName
+  const filePath = modulePath.replace(/\?export=(default|head)$/, ``)
```
Comment (LekoArts) [inline packages/gatsby/cache-dir/fast-refresh-overlay/components/runtime-errors.js:17]:
```suggestion
  // With the introduction of Metadata management the modulePath can have a resourceQuery that needs to be removed first
  const filePath = modulePath.replace(/\?export=(default|head)$/, ``)
```

File: packages/gatsby/src/utils/eslint-rules/__tests__/limited-exports-page-templates.ts [hunk 12]
```diff
@@ -78,6 +78,12 @@ describe(`no-anonymous-exports-page-templates`, () => {
       test({
         code: `import { graphql, Link } from "gatsby"\nconst Template = () => {}\nexport const query = graphql\`test\`\nexport default Template\nexports.config = () => { return ({ params }) => { defer: true }}`,
       }),
+      test({
```
Comment (LekoArts) [inline packages/gatsby/src/utils/eslint-rules/__tests__/limited-exports-page-templates.ts:81]: This test is missing cases of defining the `Head` directly in the page (it only re-exports it from somewhere). As function/arrow function etc.

See the other exports for inspiration

**Review by LekoArts — CHANGES_REQUESTED** [review]

File: packages/gatsby/src/bootstrap/requires-writer.ts [hunk 13]
```diff
@@ -248,19 +248,53 @@ const preferDefault = m => (m && m.default) || m
...
-        getAbsolutePathForVirtualModule(`$virtual`),
-        c.component
-      )
-
-      return `  "${c.componentChunkName}": () => import("${slash(
-        `./${relativeComponentPath}`
-      )}" /* webpackChunkName: "${c.componentChunkName}" */)`
-    })
-    .join(`,\n`)}
+  const asyncRequires =
```
[the remark this answers is not in the document]
Comment (marvinjude) [reply packages/gatsby/src/bootstrap/requires-writer.ts:251]: I agree that this one is quite hard to read 😀

File: packages/gatsby/src/utils/eslint-rules/__tests__/limited-exports-page-templates.ts (hunk 12 above)
Comment (marvinjude) [reply packages/gatsby/src/utils/eslint-rules/__tests__/limited-exports-page-templates.ts:81]: Good catch!

File: packages/gatsby/cache-dir/react-dom-utils.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+/* global HAS_REACT_18 */
+
+const map = new WeakMap()
+
+/**
+ * Since react 18, render and hydrate moved to react-dom/client
+ * returns correct hydrate and render function based on installed react-dom version
```
Comment (j0sh77) [inline packages/gatsby/cache-dir/react-dom-utils.js:7]: Thanks for pulling this into it's own file! 👍

File: packages/gatsby/package.json [hunk 15]
```diff
@@ -126,6 +126,7 @@
     "moment": "^2.29.1",
     "multer": "^1.4.3",
     "node-fetch": "^2.6.6",
+    "node-html-parser": "^5.3.3",
```
Comment (j0sh77) [inline packages/gatsby/package.json:129]: This should probably be in our `devDependencies`, as we don't use it outside of tests.

File: packages/gatsby/src/utils/webpack/plugins/static-query-mapper.ts [hunk 16]
```diff
@@ -14,6 +14,15 @@ import type {
 type ChunkGroup = Compilation["chunkGroups"][0]
 type EntryPoint = Compilation["asyncEntrypoints"][0]
 
+const removeExportQueryParam = (
+  path: string | undefined
+): string | undefined => {
+  if (path?.endsWith(`?export=head`) || path?.endsWith(`?export=default`)) {
```
Comment (j0sh77) [inline packages/gatsby/src/utils/webpack/plugins/static-query-mapper.ts:20]: Are we sure this is robust enough? I know `resourcePath` is compared against this below, but it seems like we either need this to outright remove all query params (and update the name of the function for future us) or support just removing what's exactly needed.

File: packages/gatsby/cache-dir/head/head-export-handler-for-browser.js [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+export function headHandlerForBrowser({
+  pageComponent,
+  staticQueryResults,
+  pageComponentProps,
+}) {
+  useEffect(() => {
+    if (pageComponent?.Head) {
+      headExportValidator(pageComponent?.Head)
+
+      const headElement = createElement(
```
Comment (j0sh77) [inline packages/gatsby/cache-dir/head/head-export-handler-for-browser.js:54]: Nit-picking: We use `createElement` a lot in this and surrounding files. Can we use traditional JSX? It's a bit easier to read in this context.

**Review by j0sh77 — COMMENTED** [review]
Excited

File: packages/gatsby/package.json (hunk 15 above)
Comment (j0sh77) [reply packages/gatsby/package.json:129]: Ignore this, just noticed this is how we grab head nodes in `cache-dir` 🤦

File: packages/gatsby/cache-dir/head/head-export-handler-for-browser.js (hunk 17 above)
Comment (marvinjude) [reply packages/gatsby/cache-dir/head/head-export-handler-for-browser.js:54]: Yeah we can do that 👍

File: e2e-tests/development-runtime/cypress/integration/head-function-export/html-insertion.js [hunk 18]
```diff
@@ -0,0 +1,99 @@
+import { page, data } from "../../../shared-data/head-function-export.js"
+
+describe(`Head function export html insertion`, () => {
+  it(`should work with static data`, () => {
+    cy.visit(page.basic)
```
[the remark this answers is not in the document]
Comment (marvinjude) [reply e2e-tests/development-runtime/cypress/integration/head-function-export/html-insertion.js]: +1

File: e2e-tests/production-runtime/cypress/integration/head-function-export/html-insertion.js [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+      .should(`equal`, data.static.base)
+    cy.getTestElement(`title`).should(`have.text`, data.static.title)
+    cy.getTestElement(`meta`)
+      .invoke(`attr`, `content`)
+      .should(`equal`, data.static.meta)
+    cy.getTestElement(`noscript`).should(`have.text`, data.static.noscript)
+    cy.getTestElement(`style`).should(`contain`, data.static.style)
+    cy.getTestElement(`link`)
+      .invoke(`attr`, `href`)
+      .should(`equal`, data.static.link)
```
[the remark this answers is not in the document]
Comment (marvinjude) [reply e2e-tests/production-runtime/cypress/integration/head-function-export/html-insertion.js:17]: That'll be quite handy! 'style', 'link' and co above are test ids though not tag names 

We could have something that asserts for text and attributes for a test id like so:

```
export const assertElementAttributesAndText = elements => {
  for (const { id, text, attributes } of elements) {
    if (text) cy.getTestElement(id).should("have.text", text)
    for (const [key, value] of Object.entries(attributes)) {
      cy.getTestElement(id).invoke("attr", key).should("equal", value)
    }
  }
}

const elements = [
  {
    id: "title",
    text: "Head function export should work",
    attributes: {
      content: "title",
    },
  },
]

assertElementAttributesAndText(elements)

```

**Review by LekoArts — APPROVED** [review]

**Pull request merged**